File: python/40-combination-sum-II.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int):
        n = len(candidates)
        candidates.sort()
        ans = []
        def dfs(i, path, target):
            if target == 0:
                ans.append(path[:])
                return
            check_dup = set()
            for j in range(i, n):
                candidate = candidates[j]
                if target-candidate < 0 or candidate in check_dup: continue
                check_dup.add(candidate)
                path.append(candidate)
                dfs(j+1, path, target-candidate)
                path.pop()
        dfs(0, [], target)
        return ans

candidates = [10,1,2,7,6,1,5]
target = 8
result = Solution().combinationSum2(candidates, target)
print(result) # [[1, 7], [1, 2, 5], [2, 6], [1, 1, 6]]

candidates = [2,5,2,1,2]
target = 5
result = Solution().combinationSum2(candidates, target)
print(result) # [[1, 2, 2], [5]]

candidates = [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
target = 27
result = Solution().combinationSum2(candidates, target)
print(result) # []

File: cpp/40. Combination Sum II.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vvi ans;
    void dfs(int i, int target, vi &path, vi &candidates) {
        int n = candidates.size();
        if(target == 0) {
            ans.push_back(path);
            return;
        }
        unordered_set<int> checkDup;
        for(int j = i; j < n; j++) {
            int candidate = candidates[j];
            if(target-candidate < 0 || checkDup.count(candidate)) continue;
            checkDup.insert(candidate);
            path.push_back(candidate);
            dfs(j+1, target-candidate, path, candidates);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vi path;
        dfs(0, target, path, candidates);
        return ans;
    }
};

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi candidates; int target; vvi ans;

    solution = Solution();
    candidates = {10,1,2,7,6,1,5};
    target = 8;
    ans = solution.combinationSum2(candidates, target);
    printAns(ans);

    solution = Solution();
    candidates = {2,5,2,1,2};
    target = 5;
    ans = solution.combinationSum2(candidates, target);
    printAns(ans);
}